File: python/packages/core/src/matimo/core/skill_registry.py
"""
Skill Registry — in-memory store and search for skills.

Mirrors: packages/core/src/core/skill-registry.ts

Provides discovery, search, filtering, and semantic search via pluggable
embeddings (TF-IDF fallback).
"""
from __future__ import annotations

import logging
from dataclasses import dataclass

from matimo.core.models import (
    SearchSkillsOptions,
    SkillContentOptions,
    SkillDefinition,
    SkillSummary,
)
from matimo.core.skill_content_parser import (
    ParsedSkillContent,
    extract_skill_content,
    list_skill_sections,
    parse_skill_sections,
)
from matimo.core.tfidf_embedding import (
    EmbeddingProvider,
    TfIdfEmbeddingProvider,
    cosine_similarity,
)
from matimo.errors import ErrorCode, MatimoError

logger = logging.getLogger("matimo")


@dataclass
class SemanticSearchResult:
    """Semantic search result with relevance score."""

    skill: SkillSummary
    score: float  # cosine similarity (0–1)


class SkillRegistry:
    """In-memory skill registry with search and semantic search."""

    def __init__(self) -> None:
        self._skills: dict[str, SkillDefinition] = {}
        self._embedding_provider: EmbeddingProvider | None = None
        self._embeddings: dict[str, list[float]] = {}
        self._embeddings_dirty: bool = True
        self._parsed_content: dict[str, ParsedSkillContent] = {}
        self._default_provider: TfIdfEmbeddingProvider | None = None

    # ------------------------------------------------------------------
    # Configuration
    # ------------------------------------------------------------------

    def set_embedding_provider(self, provider: EmbeddingProvider) -> None:
        """Set a custom embedding provider (OpenAI, Cohere, etc.)."""
        self._embedding_provider = provider
        self._embeddings_dirty = True
        logger.debug("SkillRegistry: custom embedding provider set")

    # ------------------------------------------------------------------
    # Registration
    # ------------------------------------------------------------------

    def register(self, skill: SkillDefinition) -> None:
        if not skill.name:
            raise MatimoError("Skill must have a name", ErrorCode.INVALID_SCHEMA)
        self._skills[skill.name] = skill
        self._embeddings_dirty = True
        if skill.body:
            self._parsed_content[skill.name] = parse_skill_sections(skill.body)
        logger.debug("SkillRegistry: skill registered: %s", skill.name)

    def register_all(self, skills: list[SkillDefinition]) -> None:
        for skill in skills:
            self.register(skill)

    # ------------------------------------------------------------------
    # Retrieval
    # ------------------------------------------------------------------

    def get(self, name: str) -> SkillDefinition | None:
        return self._skills.get(name)

    def get_required(self, name: str) -> SkillDefinition:
        skill = self._skills.get(name)
        if not skill:
            raise MatimoError(f"Skill not found: {name}", ErrorCode.TOOL_NOT_FOUND)
        return skill

    def get_skill_content(
        self,
        name: str,
        options: SkillContentOptions | None = None,
    ) -> str | None:
        """Get selective skill content — only the sections an agent needs."""
        parsed = self._parsed_content.get(name)
        if parsed is None:
            return None
        return extract_skill_content(parsed, options)

    def get_skill_sections(
        self,
        name: str,
    ) -> list[dict[str, object]] | None:
        """List all sections of a skill with their token costs."""
        parsed = self._parsed_content.get(name)
        if parsed is None:
            return None
        return list_skill_sections(parsed)

    # ------------------------------------------------------------------
    # Listing
    # ------------------------------------------------------------------

    def list(self) -> list[SkillSummary]:
        """List all skills (Level 1 discovery — minimal context)."""
        return [self._to_summary(s) for s in self._skills.values()]

    def get_all(self) -> list[SkillDefinition]:
        return list(self._skills.values())

    def has(self, name: str) -> bool:
        return name in self._skills

    def count(self) -> int:
        return len(self._skills)

    # ------------------------------------------------------------------
    # Search (substring + optional semantic)
    # ------------------------------------------------------------------

    def search(self, options: SearchSkillsOptions | None = None) -> list[SkillSummary]:
        if options is None:
            options = SearchSkillsOptions()

        results = list(self._skills.values())

        # Filters
        if options.category:
            cat = options.category.lower()
            results = [
                s
                for s in results
                if s.metadata and s.metadata.get("category", "").lower() == cat
            ]
        if options.difficulty:
            diff = options.difficulty.lower()
            results = [
                s
                for s in results
                if s.metadata and s.metadata.get("difficulty", "").lower() == diff
            ]
        if options.tags:
            tag_set = {t.lower() for t in options.tags}
            results = [
                s
                for s in results
                if s.metadata
                and tag_set & {
                    t.strip().lower()
                    for t in s.metadata.get("tags", "").split(",")
                }
            ]
        if options.author:
            auth = options.author.lower()
            results = [
                s
                for s in results
                if s.metadata and s.metadata.get("author", "").lower() == auth
            ]

        # Query
        if options.query:
            if options.semantic:
                scored = self._rank_by_similarity(options.query, results)
                skill_by_name = {s.name: s for s in results}
                results = [
                    skill_by_name[r.skill.name]
                    for r in sorted(scored, key=lambda x: x.score, reverse=True)
                    if r.score > 0.1 and r.skill.name in skill_by_name
                ]
            else:
                lq = options.query.lower()
                results = [
                    s
                    for s in results
                    if lq in s.name.lower() or lq in s.description.lower()
                ]

        # Pagination
        paged = results[options.offset : options.offset + options.limit]
        return [self._to_summary(s) for s in paged]

    # ------------------------------------------------------------------
    # Semantic search (async, with scores)
    # ------------------------------------------------------------------

    async def semantic_search(
        self,
        query: str,
        *,
        limit: int = 10,
        min_score: float = 0.1,
    ) -> list[SemanticSearchResult]:
        await self._ensure_embeddings()

        provider: EmbeddingProvider = self._embedding_provider or self._get_default_provider()
        query_embedding = await provider.embed(query)

        results: list[SemanticSearchResult] = []
        for name, embedding in self._embeddings.items():
            skill = self._skills.get(name)
            if not skill:
                continue
            score = cosine_similarity(query_embedding, embedding)
            if score >= min_score:
                results.append(SemanticSearchResult(skill=self._to_summary(skill), score=score))

        results.sort(key=lambda r: r.score, reverse=True)
        return results[:limit]

    # ------------------------------------------------------------------
    # Mutation
    # ------------------------------------------------------------------

    def clear(self) -> None:
        self._skills.clear()
        self._embeddings.clear()
        self._parsed_content.clear()
        self._embeddings_dirty = True

    def remove(self, name: str) -> bool:
        self._embeddings.pop(name, None)
        self._parsed_content.pop(name, None)
        self._embeddings_dirty = True
        return self._skills.pop(name, None) is not None

    # ------------------------------------------------------------------
    # Internals
    # ------------------------------------------------------------------

    async def _ensure_embeddings(self) -> None:
        if not self._embeddings_dirty:
            return

        provider: EmbeddingProvider = self._embedding_provider or self._get_default_provider()
        skills = list(self._skills.values())
        texts = [self._skill_to_text(s) for s in skills]

        if isinstance(provider, TfIdfEmbeddingProvider):
            provider.fit(texts)

        vectors = await provider.embed_batch(texts)
        self._embeddings.clear()
        for i, skill in enumerate(skills):
            self._embeddings[skill.name] = vectors[i]
        self._embeddings_dirty = False

    def _rank_by_similarity(
        self,
        query: str,
        candidates: list[SkillDefinition],
    ) -> list[SemanticSearchResult]:
        provider = self._get_default_provider()
        texts = [self._skill_to_text(s) for s in candidates]
        provider.fit([*texts, query])
        query_vec = provider.embed_sync(query)

        return [
            SemanticSearchResult(
                skill=self._to_summary(skill),
                score=cosine_similarity(query_vec, provider.embed_sync(self._skill_to_text(skill))),
            )
            for skill in candidates
        ]

    def _get_default_provider(self) -> TfIdfEmbeddingProvider:
        if self._default_provider is None:
            self._default_provider = TfIdfEmbeddingProvider()
        return self._default_provider

    @staticmethod
    def _skill_to_text(skill: SkillDefinition) -> str:
        parts = [skill.name.replace("-", " "), skill.description]
        if skill.metadata:
            if skill.metadata.get("tags"):
                parts.append(skill.metadata["tags"])
            if skill.metadata.get("category"):
                parts.append(skill.metadata["category"])
        if skill.allowed_tools:
            parts.append(" ".join(skill.allowed_tools))
        return " ".join(parts)

    @staticmethod
    def _to_summary(skill: SkillDefinition) -> SkillSummary:
        return SkillSummary(
            name=skill.name,
            description=skill.description,
            version=skill.version,
            license=skill.license,
            metadata=skill.metadata,
            source=skill.source,
        )

File: typescript/packages/core/src/core/skill-registry.ts
/**
 * Skill Registry — in-memory store and search for skills
 *
 * Mirrors ToolRegistry but for skills. Provides discovery, search, and filtering.
 * Supports both substring matching (default) and semantic search via pluggable embeddings.
 */

import {
  SkillDefinition,
  SkillSummary,
  SearchSkillsOptions,
  EmbeddingProvider,
  SkillContentOptions,
} from './types.js';
import { MatimoError, ErrorCode } from '../errors/matimo-error.js';
import { getGlobalMatimoLogger } from '../logging/index.js';
import { TfIdfEmbeddingProvider, cosineSimilarity } from './tfidf-embedding.js';
import {
  parseSkillSections,
  extractSkillContent,
  listSkillSections,
} from './skill-content-parser.js';
import type { ParsedSkillContent } from './skill-content-parser.js';

/**
 * Semantic search result with relevance score.
 */
export interface SemanticSearchResult {
  skill: SkillSummary;
  /** Cosine similarity score (0–1, higher = more relevant) */
  score: number;
}

export class SkillRegistry {
  private logger = getGlobalMatimoLogger();
  private skills = new Map<string, SkillDefinition>();

  /** Pluggable embedding provider for semantic search */
  private embeddingProvider: EmbeddingProvider | null = null;
  /** Cached embeddings: skill name → vector */
  private embeddings = new Map<string, number[]>();
  /** Whether embeddings need rebuilding */
  private embeddingsDirty = true;
  /** Cached parsed content for selective loading */
  private parsedContent = new Map<string, ParsedSkillContent>();

  /**
   * Set a custom embedding provider (OpenAI, Cohere, etc.).
   * If not set, a built-in TF-IDF provider is used as fallback.
   */
  setEmbeddingProvider(provider: EmbeddingProvider): void {
    this.embeddingProvider = provider;
    this.embeddingsDirty = true;
    this.logger.debug('SkillRegistry: custom embedding provider set');
  }

  /**
   * Register a single skill
   */
  register(skill: SkillDefinition): void {
    if (!skill.name) {
      throw new MatimoError('Skill must have a name', ErrorCode.INVALID_SCHEMA);
    }
    this.skills.set(skill.name, skill);
    this.embeddingsDirty = true;

    // Cache parsed sections for selective loading
    if (skill.body) {
      this.parsedContent.set(skill.name, parseSkillSections(skill.body));
    }

    this.logger.debug('SkillRegistry: skill registered', { name: skill.name });
  }

  /**
   * Register multiple skills
   */
  registerAll(skills: SkillDefinition[]): void {
    for (const skill of skills) {
      this.register(skill);
    }
  }

  /**
   * Get a single skill by name
   */
  get(name: string): SkillDefinition | undefined {
    return this.skills.get(name);
  }

  /**
   * Get a single skill by name (throws if not found)
   */
  getRequired(name: string): SkillDefinition {
    const skill = this.skills.get(name);
    if (!skill) {
      throw new MatimoError(`Skill not found: ${name}`, ErrorCode.TOOL_NOT_FOUND);
    }
    return skill;
  }

  /**
   * Get selective skill content — only the sections an agent needs.
   * This prevents dumping entire SKILL.md files into the LLM context.
   *
   * @example
   * // Get only the error handling section, max 500 tokens
   * registry.getSkillContent('postgres-query-operations', {
   *   sections: ['Error Handling'],
   *   maxTokens: 500,
   * })
   */
  getSkillContent(name: string, options?: SkillContentOptions): string | null {
    const parsed = this.parsedContent.get(name);
    if (!parsed) return null;
    return extractSkillContent(parsed, options);
  }

  /**
   * List all sections of a skill with their token costs.
   * Agents use this to decide which sections to load.
   */
  getSkillSections(
    name: string
  ): Array<{ path: string; level: number; tokenEstimate: number }> | null {
    const parsed = this.parsedContent.get(name);
    if (!parsed) return null;
    return listSkillSections(parsed);
  }

  /**
   * List all skills (Level 1 discovery — minimal context)
   */
  list(): SkillSummary[] {
    return Array.from(this.skills.values()).map((skill) => ({
      name: skill.name,
      description: skill.description,
      version: skill.version,
      license: skill.license,
      metadata: skill.metadata,
      source: skill.source,
    }));
  }

  /**
   * Search skills by keyword, category, difficulty, etc.
   *
   * When `options.semantic` is true, uses embedding-based similarity ranking
   * instead of substring matching. Falls back to built-in TF-IDF if no
   * external embedding provider is configured.
   */
  search(options: SearchSkillsOptions = {}): SkillSummary[] {
    const { query = '', category, difficulty, tags, author, limit = 50, offset = 0 } = options;

    let results = Array.from(this.skills.values());

    // Apply filters first (these are always exact/substring)
    if (category) {
      results = results.filter(
        (skill) => skill.metadata?.category?.toLowerCase() === category.toLowerCase()
      );
    }
    if (difficulty) {
      results = results.filter(
        (skill) => skill.metadata?.difficulty?.toLowerCase() === difficulty.toLowerCase()
      );
    }
    if (tags && tags.length > 0) {
      const metadataTags = tags.map((t) => t.toLowerCase());
      results = results.filter((skill) => {
        const skillTags = (skill.metadata?.tags || '')
          .split(',')
          .map((t) => t.trim().toLowerCase());
        return metadataTags.some((tag) => skillTags.includes(tag));
      });
    }
    if (author) {
      results = results.filter(
        (skill) => skill.metadata?.author?.toLowerCase() === author.toLowerCase()
      );
    }

    // Apply query filter (substring or semantic)
    if (query) {
      if (options.semantic) {
        // Semantic search: rank by embedding similarity
        const scored = this.rankBySimilarity(query, results);
        // Precompute a map for O(1) lookup by skill name
        const skillByName = new Map(results.map((skill) => [skill.name, skill] as const));
        results = scored
          .filter((r) => r.score > 0.1) // Minimum relevance threshold
          .sort((a, b) => b.score - a.score)
          .map((r) => skillByName.get(r.skill.name))
          .filter((skill): skill is NonNullable<typeof skill> => skill !== undefined);
      } else {
        // Substring search (original behavior)
        const lowerQuery = query.toLowerCase();
        results = results.filter((skill) => {
          const nameMatch = skill.name.toLowerCase().includes(lowerQuery);
          const descMatch = skill.description.toLowerCase().includes(lowerQuery);
          return nameMatch || descMatch;
        });
      }
    }

    // Pagination
    const paged = results.slice(offset, offset + limit);

    return paged.map((skill) => ({
      name: skill.name,
      description: skill.description,
      version: skill.version,
      license: skill.license,
      metadata: skill.metadata,
      source: skill.source,
    }));
  }

  /**
   * Semantic search with relevance scores.
   * Returns ranked results with cosine similarity scores.
   *
   * @example
   * const results = await registry.semanticSearch('How do I handle Postgres connection pooling?');
   * // → [{ skill: { name: 'postgres-query-operations', ... }, score: 0.82 }]
   */
  async semanticSearch(
    query: string,
    options: { limit?: number; minScore?: number } = {}
  ): Promise<SemanticSearchResult[]> {
    const { limit = 10, minScore = 0.1 } = options;

    await this.ensureEmbeddings();

    const provider = this.embeddingProvider || this.getDefaultProvider();
    const queryEmbedding = await provider.embed(query);

    const results: SemanticSearchResult[] = [];
    for (const [name, embedding] of this.embeddings) {
      const skill = this.skills.get(name);
      if (!skill) continue;

      const score = cosineSimilarity(queryEmbedding, embedding);
      if (score >= minScore) {
        results.push({
          skill: {
            name: skill.name,
            description: skill.description,
            version: skill.version,
            license: skill.license,
            metadata: skill.metadata,
            source: skill.source,
          },
          score,
        });
      }
    }

    return results.sort((a, b) => b.score - a.score).slice(0, limit);
  }

  /**
   * Get all skills (full definitions)
   */
  getAll(): SkillDefinition[] {
    return Array.from(this.skills.values());
  }

  /**
   * Check if a skill exists
   */
  has(name: string): boolean {
    return this.skills.has(name);
  }

  /**
   * Get the count of registered skills
   */
  count(): number {
    return this.skills.size;
  }

  /**
   * Clear all skills
   */
  clear(): void {
    this.skills.clear();
    this.embeddings.clear();
    this.parsedContent.clear();
    this.embeddingsDirty = true;
  }

  /**
   * Remove a skill
   */
  remove(name: string): boolean {
    this.embeddings.delete(name);
    this.parsedContent.delete(name);
    this.embeddingsDirty = true;
    return this.skills.delete(name);
  }

  // ─── Private: Embedding Management ──────────────────────────────────────

  /**
   * Ensure embeddings are computed and up-to-date.
   */
  private async ensureEmbeddings(): Promise<void> {
    if (!this.embeddingsDirty) return;

    const provider = this.embeddingProvider || this.getDefaultProvider();
    const skills = Array.from(this.skills.values());

    // Build text corpus: name + description + metadata tags
    const texts = skills.map((s) => this.skillToText(s));

    // If using TF-IDF, refit the vocabulary
    if (provider instanceof TfIdfEmbeddingProvider) {
      provider.fit(texts);
    }

    const vectors = await provider.embedBatch(texts);

    this.embeddings.clear();
    for (let i = 0; i < skills.length; i++) {
      this.embeddings.set(skills[i].name, vectors[i]);
    }

    this.embeddingsDirty = false;
  }

  /**
   * Synchronous ranking using TF-IDF (for the non-async search() method).
   */
  private rankBySimilarity(query: string, candidates: SkillDefinition[]): SemanticSearchResult[] {
    const provider = this.getDefaultProvider();

    // Build corpus from candidates
    const texts = candidates.map((s) => this.skillToText(s));
    provider.fit([...texts, query]); // Include query in vocab

    const queryVec = provider.embedSync(query);

    return candidates.map((skill) => {
      const skillVec = provider.embedSync(this.skillToText(skill));
      return {
        skill: {
          name: skill.name,
          description: skill.description,
          version: skill.version,
          license: skill.license,
          metadata: skill.metadata,
          source: skill.source,
        },
        score: cosineSimilarity(queryVec, skillVec),
      };
    });
  }

  /**
   * Get or create the default TF-IDF provider.
   */
  private defaultProvider: TfIdfEmbeddingProvider | null = null;
  private getDefaultProvider(): TfIdfEmbeddingProvider {
    if (!this.defaultProvider) {
      this.defaultProvider = new TfIdfEmbeddingProvider();
    }
    return this.defaultProvider;
  }

  /**
   * Convert a skill into a searchable text string for embedding.
   */
  private skillToText(skill: SkillDefinition): string {
    const parts = [skill.name.replace(/-/g, ' '), skill.description];
    if (skill.metadata?.tags) parts.push(skill.metadata.tags);
    if (skill.metadata?.category) parts.push(skill.metadata.category);
    if (skill.allowedTools) parts.push(skill.allowedTools.join(' '));
    return parts.join(' ');
  }
}
